relay-server/src/endpoints/minidump.rs

```diff
@@ -43,6 +56,67 @@ fn validate_minidump(data: &[u8]) -> Result<(), BadStoreRequest> {
     Ok(())
 }
 
+/// Convenience wrapper to let a decoder decode its full input into a buffer
+fn run_decoder(decoder: &mut Box<dyn Read>) -> Result<Vec<u8>, Error> {
+    let mut buffer = Vec::new();
+    match decoder.read_to_end(&mut buffer) {
+        Ok(_) => Ok(buffer),
+        Err(err) => Err(err),
+    }
+}
+
+/// Creates a decoder based on the magic bytes the minidump payload
+fn decoder_from(minidump_data: Bytes) -> Option<Box<dyn Read>> {
+    if minidump_data.starts_with(b"\x1F\x8B") {
+        return Some(Box::new(GzDecoder::new(Cursor::new(minidump_data))));
+    } else if minidump_data.starts_with(b"\xFD\x37\x7A\x58\x5A\x00") {
+        return Some(Box::new(XzDecoder::new(Cursor::new(minidump_data))));
+    } else if minidump_data.starts_with(b"\x42\x5A\x68") {
+        return Some(Box::new(BzDecoder::new(Cursor::new(minidump_data))));
+    } else if minidump_data.starts_with(b"\x28\xB5\x2F\xFD") {
+        return Some(Box::new(
+            ZstdDecoder::new(Cursor::new(minidump_data)).unwrap(),
+        ));
+    }
+
+    None
+}
+
+/// Tries to decode a minidump using any of the supported compression formats
+/// or returns the provided minidump payload untouched if no format where detected
+fn decode_minidump(minidump_data: Bytes) -> Result<Bytes, BadStoreRequest> {
+    match decoder_from(minidump_data.clone()) {
+        Some(mut decoder) => {
+            match run_decoder(&mut decoder) {
+                Ok(decoded) => Ok(Bytes::from(decoded)),
+                Err(err) => {
+                    // we detected a compression container but failed to decode it
+                    relay_log::trace!("invalid compression container");
+                    Err(BadStoreRequest::InvalidCompressionContainer(err))
+                }
+            }
+        }
+        None => {
+            // this means we haven't detected any compression container
+            // proceed to process the payload untouched (as a plain minidump).
+            Ok(minidump_data)
+        }
+    }
+}
+
+/// Removes any compression container file extensions from the minidump
+/// filename so it can be updated in the item. Otherwise, attachments that
+/// have been decoded would still show the extension in the UI, which is misleading.
+fn remove_container_extension(filename: String) -> String {
+    let lc_filename = filename.to_ascii_lowercase();
+    let re_find_opt = CONTAINER_EXTENSIONS_RE.find(lc_filename.as_str());
+    if let Some(re_find) = re_find_opt {
+        filename[..re_find.start()].to_owned()
+    } else {
+        filename
+    }
+}
```

Comment:
This function can be rewritten to not use a regex. We can also save some allocations if we don't account for case-sensitivity, I don't know if that's a relevant concern in practice.